### ISSUE TITLE
Problem with heatmap loss (FocalLoss)

### ISSUE THREAD

**Issue by javierlorenzod** [xingyizhou/CenterNet#841]
Hi, thanks for sharing your work.

I am trying to use the modified focal loss that you mention in your paper (the following one), but I have one problem. The loss is really big in the beginning. I look inside the function [`_neg_loss`](https://github.com/xingyizhou/CenterNet/blob/8ef87b433529ac8f8bd4f95707f6bc05052c55e9/src/lib/models/losses.py#L42) and I have found that the loss is computed as the sum of all pixels in the output.
![image](https://user-images.githubusercontent.com/6094231/97712483-cbc92e80-1abe-11eb-9ac3-495616a0bf4a.png)
In the training code, you use `loss.mean()` to obtain the mean, but loss is an scalar. Did you divide the loss by the number of pixels anywhere?

Thanks!

**Comment by xingyizhou** [comment]
Hi,
  Thanks for your interest in our work. 

 >> The loss is really big in the beginning. 

  We have two mechanisms to handle this: 1) the bias in the output heatmap is initialized to a small value (-2.19 in the paper, you can use a smaller one to make it more stable e.g., -4.6. (sigmoid(-4.6) = 0.01, sigmoid(-2.19)=0.1)). 2) we clamped the after-sigmoid value to 1e-4 to 1-1e-4.

We didn't divide the loss by number of pixels. The normalization is over the number of objects (N in the equation).

**Comment by javierlorenzod** [comment]
Ok, thanks for the detailed information. I have one last question about the loss: why do you decided to use loss without using the mean of all pixels in the output? I ask this question because I am used to see models trained with MSE loss or any other loss and at the end, they reduce it using the mean, in order to scale the norm of the gradients I think.

Thanks again!

**Comment by xingyizhou** [comment]
The true reason is we just followed CornerNet to do so. It makes sense to me because most losses come from positive pixels (= the number of objects).

**Comment by javierlorenzod** [comment]
Ok, thank you!

**Comment by Cindy0725** [comment]
> Hi, thanks for sharing your work.
> 
> I am trying to use the modified focal loss that you mention in your paper (the following one), but I have one problem. The loss is really big in the beginning. I look inside the function [`_neg_loss`](https://github.com/xingyizhou/CenterNet/blob/8ef87b433529ac8f8bd4f95707f6bc05052c55e9/src/lib/models/losses.py#L42) and I have found that the loss is computed as the sum of all pixels in the output. ![image](https://user-images.githubusercontent.com/6094231/97712483-cbc92e80-1abe-11eb-9ac3-495616a0bf4a.png) In the training code, you use `loss.mean()` to obtain the mean, but loss is an scalar. Did you divide the loss by the number of pixels anywhere?
> 
> Thanks!

Hi @javierlorenzod  I also encountered this problem. The focal loss starts from around 30000 since the code uses sum instead of mean. And it doesn't show any trend to decrease to a small value after more than 10 epochs. Did you encounter such problem? If we continue to use sum, in the normal case how much the focal loss will decrease to after the model converges? Thank you very much!
